### ISSUE TITLE
Linear regression on very simple data

### ISSUE THREAD

**Issue by burgreen** [creme-ml/creme#184]
Why does this predict `y = 0.02`? Am I wrong to expect a prediction of `y = 0.5`?

```
from creme import linear_model
from creme import preprocessing

X = []
X.append( {'x':0} )
X.append( {'x':1} )

Y = []
Y.append(0)
Y.append(1)

model = preprocessing.StandardScaler()
model |= linear_model.LinearRegression()

for i in range(2):
  model = model.fit_one( X[i], Y[i] )

y = model.predict_one( {'x':0.5} )
print( 'y =', y )
```

**Comment by raphaelsty** [comment]
@burgreen, the weights of the LinearRegression are initialized at 0 by default. 

You can initialize the weights in a more appropriate way if you have an a priori on your data.

Here I have the impression that you want a function that look like f(x) = a * x + b with a = 1 and b = 0. 

b is already initialized at 0 by default. 

To initialize the parameter a at 1: 
```
model = linear_model.LinearRegression(initializer = optim.initializers.Constant(1))
```
 Alternatively, you can also let the model initialize with the default weights and provide it with more training examples. It will provide you with more accurate results as you go along.

```python
from creme import linear_model
from creme import preprocessing
from creme import optim 

X = []
X.append( {'x':0} )
X.append( {'x':1} )

Y = []
Y.append(0)
Y.append(1)

model = linear_model.LinearRegression(initializer = optim.initializers.Constant(1))

for i in range(2):
  model = model.fit_one( X[i], Y[i] )

y = model.predict_one({'x':0.5})

print(y)
```

**Comment by burgreen** [comment]
@raphaelsty, I understand exactly what you are saying, but the machinery for the mod you have proposed is not present nor supported in the LinearRegression class. 

There are `weights` in that class that seem to provide slope information in the `_raw_dot` method. But these `weights` are overwritten by every `GLM.fit_one()` call.

**Comment by burgreen** [comment]
In the spirit of creme, the below code learns a strictly linear function, but predictions are not linear. For instance, the below code predicts `y = -1181847501.0789013`.

```
from creme import linear_model
from creme import preprocessing

model = preprocessing.StandardScaler()
model |= linear_model.LinearRegression()

for i in range(10000):
  model = model.fit_one( {'x':i}, i ) 

y = model.predict_one( {'x':0.5} )
print( 'y =', y )
```

**Comment by MaxHalford** [comment]
`LinearRegression` is trained with stochastic gradient. The reason why you do not get an output of 0.5 in your first example is because your learning is not high enough.

The `initializer` parameter @raphaelsty used will be available in the next release of `creme`.

The reason why your last example is not working is because you need to call `predict_one` in your training loop, as so:

```python
from creme import linear_model
from creme import preprocessing

model = preprocessing.StandardScaler()
model |= linear_model.LinearRegression()

for i in range(10000):
    y_pred = model.predict_one( {'x':i}) 
    model = model.fit_one( {'x':i}, i ) 

y = model.predict_one( {'x':0.5} )
print( 'y =', y )  # Prints 4397.494588297517 on my machine
```

The reason why you have to call `predict_one` is a bit tricky but I haven't found a workaround yet. Basically your model is an instance `compose.Pipeline`. The unsupervised parts of a `Pipeline` (i.e. in your case the `StandardScaler`) are updated in the `predict_one` function. If you don't call `predict_one`, the `StandardScaler` will never be updated. We do this because we want the unsupervised part of our model to learn as soon as prediction data is available, instead of waiting for the `fit_one` call. "Normally" you should always be calling `predict_one` anyway because you want to have an idea of how well you're doing. However I agree that this is unexpected behavior.

**Comment by burgreen** [comment]
@MaxHalford and @raphaelsty, thank you gentlemen for the helpful responses.

I find something like below starts generating good predictions. 
```
from creme import linear_model
from creme import preprocessing

model = preprocessing.StandardScaler()
model |= linear_model.LinearRegression()

for n in range(100):
  for i in range(100):
    y = model.predict_one( {'x':i} )
    model = model.fit_one( {'x':i}, i )

x=0.5; print( x, model.predict_one( {'x':x} ) ) # 0.5 0.395
x=1.0; print( x, model.predict_one( {'x':x} ) ) # 1 0.904
x=10.0; print( x, model.predict_one( {'x':x} ) ) # 10 9.91
x=100.0; print( x, model.predict_one( {'x':x} ) ) # 100 99.90
x=1000.0; print( x, model.predict_one( {'x':x} ) ) # 1000 952.2
```

**Comment by MaxHalford** [comment]
Yeah running multiple epochs definitely helps. You can also increase the learning as so:

```
from creme import linear_model
from creme import optim
from creme import preprocessing

model = preprocessing.StandardScaler()
model |= linear_model.LinearRegression(optimizer=optim.SGD(0.1))
```

See you around!
